[RFC] Add support for parser value_set to eBPF backend (#3)
tatry/p4c: backends/ebpf/psa/README.md

A user space application is responsible for performing periodic queries to this map to read a Digest message. It can use either
`psabpf-ctl digest get pipe`, `psabpf_digest_get_next` from psabpf C API or `bpf_map_lookup_and_delete_elem` from `libbpf` API.

### value_set

[value_set](https://p4.org/p4-spec/docs/P4-16-v1.2.2.html#sec-value-set) is a P4 lang construct allowing to determine next

> parser state based on runtime values. To achieve this compiler generates additional hash map. In select case expression

Reviewer: ```suggestion
parser state based on runtime values. The P4-eBPF compiler generates additional hash map for each `ValueSet` instance. In select case expression
```

> it is checked whether entry for given key exists in this map. Value of this map is ignored and not used. 

Reviewer: ```suggestion
Each `select()` on `ValueSet` is translated into a lookup into the BPF hash map to check if an entry for a given key exists. A value of the BPF map is ignored. 
```

> Select expression which is referred by instance of `value_set`, have to be replaced by chain of `if ... else ...` statements

Reviewer: this paragraph is too detailed IMO, can be removed

- `psa_idle_timeout` is not supported yet.
- DirectCounter and DirectMeter externs are not supported for P4 tables with implementation (ActionProfile).
- The `xdp2tc=head` mode works only for packets larger than 34 bytes (the size of Ethernet and IPv4 header).

> - `value_set` supports only exact match type and can match only single field in select expression.

Reviewer: ```suggestion
- `value_set` only supports the exact match type and can only match on a single field in the `select()` expression.
```

contribute the XDP-based version of P4-eBPF that is not fully spec-compliant, but provides higher throughput.
- **XDP2TC mode.** the generated eBPF programs use `bpf_xdp_adjust_meta()` to transfer original EtherType from XDP to TC. This mode 
may not be supported by some NIC drivers. We will add two other modes that can be used alternatively. 

> - **ValueSet support.** We plan to extend implementation to support other match kinds and more fields in select expression.

Reviewer: ```suggestion
- **ValueSet support.** We plan to extend implementation to support other match kinds and multiple fields in the `select()` expression.
```
